crates/turborepo-lib/src/turbo_json/loader.rs

```diff
@@ -0,0 +1,660 @@
+use std::collections::HashMap;
+
+use tracing::debug;
+use turbopath::{AbsoluteSystemPath, AbsoluteSystemPathBuf};
+use turborepo_errors::Spanned;
+use turborepo_repository::{
+    package_graph::{PackageInfo, PackageName},
+    package_json::PackageJson,
+};
+
+use super::{Pipeline, RawTaskDefinition, TurboJson, CONFIG_FILE};
+use crate::{
+    config::Error,
+    run::{task_access::TASK_ACCESS_CONFIG_PATH, task_id::TaskName},
+};
+
+/// Structure for loading TurboJson structures.
+/// Depending on the strategy used, TurboJson might not correspond to
+/// `turbo.json` file.
+#[derive(Debug, Clone)]
+pub struct TurboJsonLoader {
+    repo_root: AbsoluteSystemPathBuf,
+    cache: HashMap<PackageName, TurboJson>,
+    strategy: Strategy,
+}
+
+#[derive(Debug, Clone)]
+enum Strategy {
+    SinglePackage {
+        root_turbo_json: AbsoluteSystemPathBuf,
+        package_json: PackageJson,
+    },
+    Workspace {
+        // Map of package names to their package specific turbo.json
+        packages: HashMap<PackageName, AbsoluteSystemPathBuf>,
+    },
+    WorkspaceNoTurboJson {
+        // Map of package names to their scripts
+        packages: HashMap<PackageName, Vec<String>>,
+    },
+    TaskAccess {
+        root_turbo_json: AbsoluteSystemPathBuf,
+        package_json: PackageJson,
+    },
+    Noop,
+}
+
+impl TurboJsonLoader {
+    /// Create a loader that will load turbo.json files throughout the workspace
+    pub fn workspace(
+        repo_root: AbsoluteSystemPathBuf,
+        packages: HashMap<PackageName, AbsoluteSystemPathBuf>,
```

Comment:
This could be made more ergonomic by just taking `&PackageGraph` and keeping the extraction logic in this module. I avoided this to allow for testing loaders without constructing an entire package graph.